filebeat/scripts/tester/main.go

```diff
@@ -0,0 +1,333 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/elastic/beats/filebeat/harvester/encoding"
+	"github.com/elastic/beats/filebeat/harvester/reader"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/match"
+)
+
+func main() {
+	esURL := flag.String("elasticsearch", "http://localhost:9200", "Elasticsearch URL")
+	path := flag.String("pipeline", "", "Path to pipeline")
+	modulesPath := flag.String("modules", "./modules", "Path to modules")
+
+	log := flag.String("log", "", "Single log line to test")
+	logfile := flag.String("logfile", "", "Path to log file")
+
+	multiPattern := flag.String("multiline.pattern", "", "Multiline pattern")
+	multiNegate := flag.Bool("multiline.negate", false, "Multiline negate")
+	multiMode := flag.String("multiline.mode", "before", "Multiline mode")
+	maxBytes := flag.Int("maxbytes", 10485760, "Number of max bytes to be read")
+
+	verbose := flag.Bool("verbose", false, "Call Simulate API with verbose option")
+	simulateVerbose := flag.Bool("simulate.verbose", false, "Print full output of Simulate API with verbose option")
+	flag.Parse()
+
+	if *path == "" {
+		os.Stderr.WriteString("Error: -pipeline is required\n")
+		os.Exit(1)
+	}
+
+	if *log == "" && *logfile == "" {
+		os.Stderr.WriteString("Error: -log or -logs has to be specified\n")
+		os.Exit(1)
+	}
+
+	if *multiPattern != "" && *logfile == "" {
+		os.Stderr.WriteString("Error: -multiline-pattern is set but -logfile is not\n")
+		os.Exit(1)
+	}
+
+	var logs []string
+	var err error
+	if *logfile != "" {
+		logs, err = getLogsFromFile(*logfile, *multiPattern, *multiNegate, *multiMode, *maxBytes)
+		if err != nil {
+			os.Stderr.WriteString(fmt.Sprintf("Error while reading logs from file: %v\n", err))
+			os.Exit(2)
+		}
+	} else {
+		logs = []string{*log}
+	}
+
+	paths, err := getPipelinePath(*path, *modulesPath)
+	if err != nil {
+		os.Stderr.WriteString(err.Error())
+		os.Exit(3)
+	}
+	if len(paths) == 0 {
+		os.Stderr.WriteString("No pipeline file was found\n")
+		os.Exit(3)
+	}
+
+	for _, path := range paths {
+		err = testPipeline(*esURL, path, logs, *verbose, *simulateVerbose)
+		if err != nil {
+			os.Stderr.WriteString(err.Error())
+			os.Exit(4)
+		}
+	}
+}
+func getLogsFromFile(logfile, multiPattern string, multiNegate bool, matchMode string, maxBytes int) ([]string, error) {
+	f, err := os.Open(logfile)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	encFactory, ok := encoding.FindEncoding("utf8")
```

Comment:
encoding should be configurable. Especially windows users will oftentimes have utf16